### ISSUE TITLE
Better Docker build image

### ISSUE THREAD

**Issue by xinbinhuang** [fishtown-analytics/dbt#2498]
### Describe the feature
Move the inline Dockerfile to a separate `Dockerfile` 
https://github.com/fishtown-analytics/dbt/blob/dev/octavius-catto/scripts/build-dbt.py#L717-L745

Relevant: #2457 

#DOCKER

### Describe alternatives you've considered

Something like this. and it is more easy to tune and optimize.

When we need to build the image, pass those arguments with `--build-arg`, and we don't need to keep a list of Dockerfiles.

```bash
ARG REQUIREMENTS_PATH 
ARG REQUIREMENTS_FILE_NAME
ARG DIST_PATH
ARG WHEEL_PATH
ARG PYTHON_BASE_IMAGE="python:3.6-slim-buster"

FROM ${PYTHON_BASE_IMAGE}
RUN apt-get update && \
    apt-get install -y  --no-install-recommends \
        git software-properties-common make build-essential \
        ca-certificates libpq-dev && \
    apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
COPY $REQUIREMENTS_PATH ./$REQUIREMENTS_FILE_NAME
COPY $DIST_PATH ./dist
RUN pip install --upgrade pip setuptools
RUN pip install --requirement ./$REQUIREMENTS_FILE_NAME
RUN pip install $WHEEL_PATH
RUN useradd -mU dbt_user
ENV PYTHONIOENCODING=utf-8
ENV LANG C.UTF-8
WORKDIR /usr/app
VOLUME /usr/app
USER dbt_user
ENTRYPOINT dbt
```

### Additional context
Is this feature database-specific? Which database(s) is/are relevant? Please include any other relevant context here.

### Who will this benefit?

Everyone. This helps 
- maintain a better Docker image - easier to extend etc 
- makes it easier to optimize in the future. 
- Multi Python version images
- This can be further configured to integrate with the CI images with Multi-Stage build

**Comment by xinbinhuang** [comment]
I can help create a PR on this if you guys agree

**Comment by beckjake** [comment]
Thanks for the feedback! I'll definitely keep this in mind for #2458 (which I think is more of the "right place" than #2457). 

I don't think it's likely that this will happen for 0.17.0, but I would expect this to be in 0.18.0 if we don't hit any problems with it.

**Comment by xinbinhuang** [comment]
Oh.. it should be #2458 indeed, sorry for the typo.

Are you open to a PR from me? I can first move the inline textwrap Dockerfile to the root directory.

**Comment by beckjake** [comment]
@xinbinhuang Sure! Instead of the root directory, could it go into the `docker` folder? We already have a `Dockerfile` in the root directory for the test container.

**Comment by xinbinhuang** [comment]
Hi @beckjake , I took a look at the codebase, and I realize that it may involve rewriting the whole logic of `DockerBuilder` and how you store and push docker images. So I want to discuss it here first to see if it it's the right moment to do so alongside with other tasks that you are doing in the project.

So the eventual goal would be to have a single `Dockerfile` only and use `ARG` to provide builds for different releases & python versions & with different extra dependencies etc. These images will be built and then pushed to Docker Hub in the CI. So at then end, there will be different images on Docker hub and only one image at the repo. The benefit is that we treat Dockerfile as a template, and it can be easily extended and fit into different builds.

Another thing is to replace the docker related `run_command` with the official [Python SDK](https://docker-py.readthedocs.io/en/stable/). This make the supply of `--build-arg` easier.

What do you think?

Best
Bin

**Comment by xinbinhuang** [comment]
https://github.com/fishtown-analytics/dbt/blob/05634e61a70f4977726acbf05232a63899e62442/scripts/build-dbt.py#L763-L773

would become something like this

```python
 def create_docker_image(self): 
     ....

     self.client.images.build(
            path=self.dockerfile_path,
            tag=self.image_tag,
            buildargs={
                  'REQUIREMENTS_PATH'=requirements_path,
                  'REQUIREMENTS_FILE_NAME'=self.requirements_file_name,
                  'DIST_PATH'=dist_path,
                  'WHEEL_PATH'=wheel_paths,
                  'PYTHON_BASE_IMAGE'="python:3.8.1-slim-buster"
            }
     )
````

**Comment by beckjake** [comment]
Hey @xinbinhuang - I've actually started on this on my own over here in our new dbt-release repository: https://github.com/fishtown-analytics/dbt-release/blob/master/Dockerfile

In that repository I'm actually planning to use a github action from docker itself to build the image. Longer-term we likely won't be maintaining the release Dockerfile or the build-dbt script in the dbt repository.

**Comment by johnoscott** [comment]
The current Dockerfile uses 

```FROM ubuntu:18.04```

but this change is proposing 

```FROM ${PYTHON_BASE_IMAGE}```

is the plan to use an official python docker image like `python:3.8.3-slim-buster` ?

**Comment by xinbinhuang** [comment]
> The current Dockerfile uses
> 
> `FROM ubuntu:18.04`
> 
> but this change is proposing
> 
> `FROM ${PYTHON_BASE_IMAGE}`
> 
> is the plan to use an official python docker image like `python:3.8.3-slim-buster` ?

That will be just a placeholder. You can easily substitute whatever base image you like with `--build-arg PYTHON_BASE_IMAGE=<some-image>`.

**Comment by xinbinhuang** [comment]
@beckjake I see. I guess then this issue can be closed for now as it is not relevant to this repo. 

That being said, what do you feel about the change, and do you want me to submit a new PR in the release repo?

**Comment by beckjake** [comment]
@xinbinhuang I believe I did implement that change! The new Dockerfile takes the things it cares about as buildArgs.

@johnoscott The Ubuntu dockerfile you're looking at is our test env Dockerfile. The "official" dockerfile that gets pushed to `fishtownanalytics/dbt` has always been a `3.8.x-slim-buster`-based image (I think)

**Comment by johnoscott** [comment]
OK good to know. Thanks for the clarification.

**Comment by jtcohen6** [comment]
I believe this was resolved by #3019
